swift/ql/lib/codeql/swift/controlflow/internal/AstControlFlowTrees.qll

```diff
@@ -0,0 +1,264 @@
+private import swift
+private import Completion
+private import ControlFlowGraphImplShared
+private import ControlFlowElements
+
+module ControlFlowTrees {
+  abstract class AstControlFlowTree extends AstNode {
+    /** Holds if `first` is the first element executed within this element. */
+    pragma[nomagic]
+    abstract predicate first(ControlFlowElement first);
+
+    /**
+     * Holds if `last` with completion `c` is a potential last element executed
+     * within this element.
+     */
+    pragma[nomagic]
+    abstract predicate last(ControlFlowElement last, Completion c);
+
+    /** Holds if abnormal execution of `child` should propagate upwards. */
+    abstract predicate propagatesAbnormal(ControlFlowElement child);
+
+    /**
+     * Holds if `succ` is a control flow successor for `pred`, given that `pred`
+     * finishes with completion `c`.
+     */
+    pragma[nomagic]
+    abstract predicate succ(ControlFlowElement pred, ControlFlowElement succ, Completion c);
+  }
+
+  private class ControlFlowTreeAdaptor extends ControlFlowTree {
+    AstControlFlowTree tree;
+
+    ControlFlowTreeAdaptor() { tree = this.asAstNode() }
+
+    override predicate first(ControlFlowElement first) { tree.first(first) }
+
+    override predicate last(ControlFlowElement last, Completion c) { tree.last(last, c) }
+
+    override predicate propagatesAbnormal(ControlFlowElement child) {
+      tree.propagatesAbnormal(child)
+    }
+
+    override predicate succ(ControlFlowElement pred, ControlFlowElement succ, Completion c) {
+      tree.succ(pred, succ, c)
+    }
+  }
+}
+
+import ControlFlowTrees
+
+/**
+ * Holds if `first` is the first element executed within control flow
+ * element `cft`.
+ */
+predicate astFirst(AstControlFlowTree cft, ControlFlowElement first) { cft.first(first) }
```

Comment:
It means the same (now), but I prefer `first(any(ControlFlowTree tree | tree.asAstNode()), first)`.